File: clients/Java/src/main/java/model/DebugCommand.java
package model;

import util.StreamUtil;

public abstract class DebugCommand {
    public abstract void writeTo(java.io.OutputStream stream) throws java.io.IOException;
    public static DebugCommand readFrom(java.io.InputStream stream) throws java.io.IOException {
        switch (StreamUtil.readInt(stream)) {
            case Add.TAG:
                return Add.readFrom(stream);
            case Clear.TAG:
                return Clear.readFrom(stream);
            default:
                throw new java.io.IOException("Unexpected tag value");
        }
    }

    public static class Add extends DebugCommand {
        public static final int TAG = 0;
        private model.DebugData data;
        public model.DebugData getData() { return data; }
        public void setData(model.DebugData data) { this.data = data; }
        public Add() {}
        public Add(model.DebugData data) {
            this.data = data;
        }
        public static Add readFrom(java.io.InputStream stream) throws java.io.IOException {
            Add result = new Add();
            result.data = model.DebugData.readFrom(stream);
            return result;
        }
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
            data.writeTo(stream);
        }
    }

    public static class Clear extends DebugCommand {
        public static final int TAG = 1;
        public Clear() {}
        public static Clear readFrom(java.io.InputStream stream) throws java.io.IOException {
            Clear result = new Clear();
            return result;
        }
        @Override
        public void writeTo(java.io.OutputStream stream) throws java.io.IOException {
            StreamUtil.writeInt(stream, TAG);
        }
    }
}

File: clients/Scala/src/main/scala/model/DebugCommand.scala
package model

import util.StreamUtil

sealed trait DebugCommand {
    def writeTo(stream: java.io.OutputStream)
}
object DebugCommand {
    case class Add(data: model.DebugData) extends DebugCommand {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, Add.TAG)
            data.writeTo(stream)
        }
    }
    object Add {
        val TAG: Int = 0
        def readFrom(stream: java.io.InputStream): Add = Add(
            model.DebugData.readFrom(stream)
            )
    }

    case class Clear() extends DebugCommand {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, Clear.TAG)
        }
    }
    object Clear {
        val TAG: Int = 1
        def readFrom(stream: java.io.InputStream): Clear = Clear(
            )
    }

    case class SetAutoFlush(enable: Boolean) extends DebugCommand {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, SetAutoFlush.TAG)
            StreamUtil.writeBoolean(stream, enable)
        }
    }
    object SetAutoFlush {
        val TAG: Int = 2
        def readFrom(stream: java.io.InputStream): SetAutoFlush = SetAutoFlush(
            StreamUtil.readBoolean(stream)
            )
    }

    case class Flush() extends DebugCommand {
        override def writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, Flush.TAG)
        }
    }
    object Flush {
        val TAG: Int = 3
        def readFrom(stream: java.io.InputStream): Flush = Flush(
            )
    }

    def readFrom(stream: java.io.InputStream): DebugCommand = {
        StreamUtil.readInt(stream) match {
            case Add.TAG => Add.readFrom(stream)
            case Clear.TAG => Clear.readFrom(stream)
            case SetAutoFlush.TAG => SetAutoFlush.readFrom(stream)
            case Flush.TAG => Flush.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
    }
}
